client/src/main/scala/org/http4s/netty/client/Http4sHandler.scala

```diff
@@ -29,33 +29,70 @@ import org.http4s.Response
 import org.http4s.netty.NettyModelConversion
 
 import java.io.IOException
+import java.nio.channels.ClosedChannelException
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
 
-private[netty] class Http4sHandler[F[_]](cb: Http4sHandler.CB[F], dispatcher: Dispatcher[F])(
-    implicit F: Async[F])
+private[netty] class Http4sHandler[F[_]](dispatcher: Dispatcher[F])(implicit F: Async[F])
     extends ChannelInboundHandlerAdapter {
 
   private[this] val logger = org.log4s.getLogger
-  val modelConversion = new NettyModelConversion[F]
+  private val modelConversion = new NettyModelConversion[F]
+  private val promises =
+    collection.mutable.Queue[Either[Throwable, Resource[F, Response[F]]] => Unit]()
+  // By using the Netty event loop assigned to this channel we get two benefits:
+  //  1. We can avoid the necessary hopping around of threads since Netty pipelines will
+  //     only pass events up and down from within the event loop to which it is assigned.
+  //     That means calls to ctx.read(), and ct.write(..), would have to be trampolined otherwise.
+  //  2. We get serialization of execution: the EventLoop is a serial execution queue so
+  //     we can rest easy knowing that no two events will be executed in parallel.
+  private[this] var eventLoopContext: ExecutionContext = _
+
+  private var pending: Future[Unit] = Future.unit
+
+  private[netty] def addCallback(
+      callback: Either[Throwable, Resource[F, Response[F]]] => Unit): Unit =
+    void(promises.enqueue(callback))
```

Comment:
I really like where this is going.
As it is it's not thread safe (the netty thread and the app-space thread will be mutating `promises`). It would also be cool if we could make this perform the whole dispatch, eg `def dispatch(callback, request)` since we're already doing half the dispatch here anyway.